### DISCUSSION TITLE
Support request for ST7701S 320x960 4.58in IPS RGB display

### DISCUSSION THREAD

**Discussion by nikthefix** [moononournation/Arduino_GFX#267] [Ideas]
Support request for ST7701S 320x960 4.58in IPS.
I'm close to getting it working but can't find the timing data.

Thanks

Nik

**Comment by rjchasse** [comment]
Have you gotten this working? I'm also interested in using this display but can't seem to find support for it.

**Comment by nikthefix** [comment]
@rjchasse 
Hello, 

Yes I think I got it working in the end although I had to guess at the timings. I've seen many more distributors of this display on Aliexpress now so perhaps one of them would supply the timing data (with a purchase naturally). Have you ordered one or just thinking of doing so?
I never found the proper documentation.

I'll dig mine out and revisit it - glad you reminded me!

With the init sequence and timings the rest should be easy and well supported by Arduino_GFX and Lovyan_GFX. Last time I looked Arduino_GFX didn't have RGB support for esp32-s3 with arbitrary pin assignments. That's probably changed now. Both libs offer good ST7701 examples.

Additional: If you're interested to use this form factor for a car display or some application where the long landscape aspect ratio is appealing then you might consider a display with internal gram. In my experience the ESP32-S3 RGB LCD peripheral is hard to make work well above 400x400px with a decent frame rate. Competition to access PSRAM leads to glitches unless you design around the demand. Frustratingly, as your application expands so potentially do the glitches as the PSRAM fetch of the frame buffer must not be interrupted yet there are processes you might add which may take higher priority. Now there lots of ways around this issue and you can find much info on the Espressif forums.

But I recon QSPI or MCU 16bit parallel is a better way to go. These always seem to have internal buffers.

Additional2: Many of these long thin displays are native portrait (why I can't imagine). If they don't offer a hardware rotate / scan direction orientation then you'll be forever paying the price of the overhead of rotating your gui in software. The datasheets don't help as they often show broad functionality for the driver IC which has not been fully implemented in the display hardware. You just have to suck it and see.

I think makers are attracted to these RGB displays because they're cheap and come in a wide range of fascinating form factors. I'm not sure that they're the best fit for ESP32-S3, especially if you want to be able to do other things at the same time.

nik

**Reply by rjchasse** [reply]
Thanks for the response. My plan for using this display is to display a string of text from an MQTT string. Nothing fancy. I am using a board from Adafruit called the Qualia S3 RGB-666 and I believe the Arduino_GFX library supports the board but not this particular display. Please see this post in the Adafruit forum where I provided my best shot at adding the timing and init sequence. The code compiles and uploads successfully and I even get serial messages (with seemingly promising results) however the display remains black with the backlight on.

https://forums.adafruit.com/viewtopic.php?t=206598

Please take a look if possible and let me know if the timing, setup, and init sequence look okay or if I am missing something.

Thanks,
Ronnie

**Comment by nikthefix** [comment]
OK, that's encouraging. Did you try any other RGB displays like the kind adafruit use in their Qualia demos?

**Reply by rjchasse** [reply]
No I haven't tried other displays. The other sizes and shapes aren't going to work for my use case, unfortunately.

**Comment by nikthefix** [comment]
If you can duplicate their setup with maybe the 2.1" round display. Get it up and running to your satisfaction. Substitute it for your panoramic display (40-pin right?). Resize buffers to accommodate the new pixel count. Change timings for the new display. The init codes may be different but without the datasheet who knows. They're likely to be the same. They were when I tried it.

Start with something you know works. Then make incremental changes. Swapping out the display is hardly an incremental change but at least you know it uses the same driver and the pin layout is the same (in this case).

You have to start with something that works otherwise there are simply too many blind alleys to choose from.

**Reply by rjchasse** [reply]
I've definitely tried that already. Using the PDQgraphicstest.ino I have tried using many different ST7701S options from Arduino_GFX_dev_device.h and none of them worked. I even tried making my own st7701_typeX_init_operations for this bar display according to Adafruits own "Init Code Listing" in the technical details of this page https://www.adafruit.com/product/5805. 

This is the code I used. I created it based on the examples for other displays but I added what I thought were the appropriate details.
```
Arduino_ESP32RGBPanel *rgbpanel = new Arduino_ESP32RGBPanel(
    13 /* DE */, 14 /* VSYNC */, 15 /* HSYNC */, 12 /* PCLK */,
    16 /* R0 */, 17 /* R1 */, 18 /* R2 */, 19 /* R3 */, 20 /* R4 */,
    8 /* G0 */, 9 /* G1 */, 10 /* G2 */, 11 /* G3 */, 12 /* G4 */, 13 /* G5 */,
    0 /* B0 */, 1 /* B1 */, 2 /* B2 */, 3 /* B3 */, 4 /* B4 */,
    1 /* hsync_polarity */, 30 /* hsync_front_porch */, 10 /* hsync_pulse_width */, 50 /* hsync_back_porch */,
    1 /* vsync_polarity */, 15 /* vsync_front_porch */, 2 /* vsync_pulse_width */, 17 /* vsync_back_porch */);
Arduino_RGB_Display *gfx = new Arduino_RGB_Display(
    320 /* width */, 960 /* height */, rgbpanel, 0 /* rotation */, true /* auto_flush */,
    bus, GFX_NOT_DEFINED /* RST */, st7701_type10_init_operations, sizeof(st7701_type10_init_operations));
```

Here is also the st7701_type10_init_operations I used:
```
static const uint8_t st7701_type10_init_operations[] = {
    BEGIN_WRITE,
    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 
    0x77, 0x01, 0x00, 0x00, 0x13,

    WRITE_C8_D8, 0xEF, 0x08,

    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 
    0x77, 0x01, 0x00, 0x00, 0x10,

    WRITE_C8_D16, 0xC0, 0x77, 0x00,
    WRITE_C8_D16, 0xC1, 0x09, 0x08,
    WRITE_C8_D16, 0xC2, 0x01, 0x02,

    WRITE_C8_D8, 0xC3, 0x02,
    WRITE_C8_D8, 0xCC, 0x10,

    WRITE_COMMAND_8, 0xB0, // Positive Voltage Gamma Control
    WRITE_BYTES, 16,
    0x40, 0x14, 0x59, 0x10,
    0x12, 0x08, 0x03, 0x09,
    0x05, 0x1E, 0x05, 0x14,
    0x10, 0x68, 0x33, 0x15,

    WRITE_COMMAND_8, 0xB1, // Negative Voltage Gamma Control
    WRITE_BYTES, 16,
    0x40, 0x08, 0x53, 0x09,
    0x11, 0x09, 0x02, 0x07,
    0x09, 0x1A, 0x04, 0x12,
    0x12, 0x64, 0x29, 0x29,

    // PAGE1
    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 
    0x77, 0x01, 0x00, 0x00, 0x11,

    WRITE_C8_D8, 0xB0, 0x6D, // Vop=4.7375v
    WRITE_C8_D8, 0xB1, 0x1D, // VCOM=32
    WRITE_C8_D8, 0xB2, 0x87, // VGH=15v
    WRITE_C8_D8, 0xB3, 0x80,
    WRITE_C8_D8, 0xB5, 0x49, // VGL=-10.17v
    WRITE_C8_D8, 0xB7, 0x85,
    WRITE_C8_D8, 0xB8, 0x20, // AVDD=6.6 & AVCL=-4.6
    WRITE_C8_D8, 0xC1, 0x78,
    WRITE_C8_D8, 0xC2, 0x78,
    WRITE_C8_D8, 0xD0, 0x88,

    WRITE_COMMAND_8, 0xE0,
    WRITE_BYTES, 3, 
    0x00, 0x00, 0x02,

    WRITE_COMMAND_8, 0xE1,
    WRITE_BYTES, 11,
    0x02, 0x8C, 0x00, 0x00,
    0x03, 0x8C, 0x00, 0x00,
    0x00, 0x33, 0x33,

    WRITE_COMMAND_8, 0xE2,
    WRITE_BYTES, 13,
    0x33, 0x33, 0x33, 0x33,
    0xC9, 0x3C, 0x00, 0x00,
    0xCA, 0x3C, 0x00, 0x00,
    0x00,

    WRITE_COMMAND_8, 0xE3,
    WRITE_BYTES, 4, 
    0x00, 0x00, 0x33, 0x33,

    WRITE_C8_D16, 0xE4, 0x44, 0x44,

    WRITE_COMMAND_8, 0xE5,
    WRITE_BYTES, 16,
    0x05, 0xCD, 0x82, 0x82,
    0x01, 0xC9, 0x82, 0x82,
    0x07, 0xCF, 0x82, 0x82,
    0x03, 0xCB, 0x82, 0x82,

    WRITE_COMMAND_8, 0xE6,
    WRITE_BYTES, 4, 
    0x00, 0x00, 0x33, 0x33,

    WRITE_C8_D16, 0xE7, 0x44, 0x44,

    WRITE_COMMAND_8, 0xE8,
    WRITE_BYTES, 16,
    0x06, 0xCE, 0x82, 0x82,
    0x02, 0xCA, 0x82, 0x82,
    0x08, 0xD0, 0x82, 0x82,
    0x04, 0xCC, 0x82, 0x82,

    WRITE_COMMAND_8, 0xEB,
    WRITE_BYTES, 7,
    0x08, 0x01, 0xE4, 0xE4,
    0x88, 0x00, 0x40,

    WRITE_COMMAND_8, 0xEC, 
    WRITE_BYTES, 3,
    0x00, 0x00, 0x00,

    WRITE_COMMAND_8, 0xED,
    WRITE_BYTES, 16,
    0xFF, 0xF0, 0x07, 0x65,
    0x4F, 0xFC, 0xC2, 0x2F,
    0xF2, 0x2C, 0xCF, 0xF4,
    0x56, 0x70, 0x0F, 0xFF,

    //-----------VAP & VAN---------------
    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 4, 
    0x77, 0x01, 0x00, 0x00, 0x00,

    WRITE_COMMAND_8, 0x11, // Sleep Out
    END_WRITE,

    DELAY, 120,

    BEGIN_WRITE,
    WRITE_C8_D8, 0x35, 0x00,
    WRITE_C8_D8, 0x3A, 0x66,

    WRITE_COMMAND_8, 0x11, // Sleep Out
    WRITE_COMMAND_8, 0x29, // Display On
    END_WRITE};
```

**Reply by jerrylong-km** [reply]
I have used your init code on my GC9503 dispaly(6inch, 360x960), it shows some color line. How can I adjust init code?
![9](https://github.com/moononournation/Arduino_GFX/assets/53472123/df47d1bb-4f35-41a0-891e-2a71f895154d)

**Comment by nikthefix** [comment]
OK but have you had it working with a display / sketch from which your project is derived?

**Reply by rjchasse** [reply]
No I haven't. I don't have another display to work with. I'm of course just assuming the code for the supported displays works, so I'm trying to adapt it for my display. Adsfruit provided the timing diagram and data sheet. Thats where I tried to find those numbers in the code for R G B Hsync Vsync etc.

**Reply by nikthefix** [reply]
Hsync and Vsync are critical and vary from display to display along with the permissible data rates. I'll try to find what I used and post it here. Do yourself a favor and get it working on a 400x400px  ST7701S first.  It's a short step then to all the others but you can't assume that they all follow suit. There be wolves in the datasheet.

**Reply by rjchasse** [reply]
Thanks! So do you think it is likely not the init commands, but more likely the timing and rgb params? I noticed a couple of the parameters printed on the PCB of the board. Do you think these are accurate? I thought those were display specific, not board specific...

See the picture attached. 
![image](https://github.com/moononournation/Arduino_GFX/assets/24484821/1406347b-44b5-42b3-9ad5-5d462f0fd37e)

**Reply by nikthefix** [reply]
Those parameters do relate to the actual display but are only valid for the selection of RGB displays that adafruit have tried and recommend. Although the pins should be the same the timings will likely be different for your chosen display model as it's not on the list - so to speak. The good news is that if you experiment with the timings, when you get close you should see 'something'. Reduce the parallel data rate and the spi setup data rate to something lower than in the example code - but not too much - maybe half. Confirm that your display has 40 pins. There is a 45 pin version which looks almost identical and still fits the ffc connector. This wont work of course. 

Does the backlight come on when you power up using the adafruit example sketch?

nik
